workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/ListenToCoordinates.kt

```diff
@@ -0,0 +1,101 @@
+package com.squareup.workflow1.ui.container
+
+import android.graphics.Rect
+import android.view.View
+import android.view.View.OnAttachStateChangeListener
+import android.view.ViewTreeObserver
+import androidx.core.view.doOnAttach
+
+/**
+ * Calls [block] with the absolute screen coordinates of the receiving [View] while
+ * it is attached, updating each time they change. If the [View] is attached already,
+ * the first call to [block] is made immediately.
+ *
+ * Callbacks will cease once the view is detached or the returned "cancel handler" is invoked.
+ *
+ * Case: view is attached.
+ *     listenToCoordinates -> block()
+ *       view changes position -> block()
+ *       view changes position -> block()
+ *     cancel handler -> no more listening.
+ *
+ * Case: view is detached.
+ *     listenToCoordinates
+ *       view is attached -> block()
+ *       view changes position -> block()
+ *       view changes position -> block()
+ *       view is detached -> no more listening.
+ *
+ * @param block lambda to execute for the current position and each future change.
+ * @return a lambda that can be invoked to cancel the tracking.
+ */
+internal fun View.listenToCoordinates(
+  block: (Rect) -> Unit
+): () -> Unit {
```

Comment:
Couldn't we throw a simple alias on this so it is more identifiable as a `Canceller`?

```
typealias Cancel = () -> Unit
```

Reply:
Yeah, I think this API does need work. But I don't actually use it at all in this PR, should hold off on refining it until it's actually driving something.

Hmm. As I type this, I'm thinking that `RectProvider` is actually redundant with the `ViewEnvironment`. In the example from my reply to @helios175, it seems like the container could just provide the cover area and target area hints that way. It would still use `listenToCoordinates` to know when to update the rect, but there's no benefit to building that callback API directly into `DialogFactory`.

I should really rebase this PR to remove `ListenToCoordinates` and `RectProvider`, and introduce them with the next PR when/if they're put into action.